packages/fxa-auth-server/scripts/audit-tokens.ts

```diff
@@ -0,0 +1,519 @@
+function buildLimit(table: string) {
+  const sampleSize = getSampleSize(table);
+  if (sampleSize) {
+    return `LIMIT ${sampleSize}`;
+  }
+  return '';
+}
+
+/** Adds the number of rows sampled and the total table size to each result set. This can be useful metadata when graphing metrics. */
+function decorateResultWithTableStats(table: string, result: any) {
+  const tableSize = getTableSize(table);
+  let sampleSize = getSampleSize(table);
+  if (tableSize < sampleSize) {
+    sampleSize = tableSize;
+  }
+  result.table_size = tableSize;
+  result.sample_size = sampleSize;
+}
+
+/** Conducts audit with a sql query that outputs a single row stats. */
+async function audit(name: string, raw: string) {
+  // Make sure query passes filter. We always run the RowCount tests since they
+  // are needed by other audits.
+  const filter = program.grep ? new RegExp(program.grep) : undefined;
+  const skip = !/RowCount/i.test(name) && filter && !filter.test(name);
+  if (skip) {
+    log.info('audit', { msg: `-- Excluding ${name} due to grep filter.` });
+
+    if (program.verbose) {
+      console.log(`-- Excluding ${name} due to grep filter.`);
+    }
+
+    return;
+  } else {
+    if (program.verbose) {
+      console.log('!!! ', name);
+    }
+  }
+
+  if (program.dry) {
+    logResult(name, raw, '');
+    return '';
+  }
+
+  try {
+    const isolationLevel = 'read uncommitted';
+    const trx = await knex.transaction({ isolationLevel });
+    const rawResult = await trx.raw(raw);
+    await trx.commit();
+    return formatResult(rawResult);
+  } catch (err) {
+    log.error(err);
+    if (program.verbose) {
+      console.log(err);
+    }
+  }
+}
+
+/* Holds on to row counts per table. These are useful for other statistics. */
+const rowCounts: Map<string, number> = new Map();
+
+/** Queries for current row counts. */
+export async function auditRowCounts(table: string) {
+  function buildQuery(table: string) {
+    return `
+  SELECT
+    table_rows AS table_size
+  FROM INFORMATION_SCHEMA.TABLES
+  WHERE table_schema = '${table.split('.')[0]}' and table_name = '${
+      table.split('.')[1]
+    }'
+  `;
+  }
+
+  const name = `${table}.RowCount.`;
+  const query = buildQuery(table);
+  let result = await audit(name, query);
+
+  logResult(name, query, result);
+  emitStats(name, result);
+
+  if (typeof result.table_size === 'number') {
+    rowCounts.set(table, result?.table_size);
+  }
+
+  return result;
+}
+
+/** Groups tables by year month and gets counts. */
+export async function auditAge(
+  table: string,
+  colName: string,
+  colSort: string
+) {
+  function buildQuery(table: string, timeCol: string, sortCol: string) {
+    return `
+  SELECT
+    DATE_FORMAT(FROM_UNIXTIME(${timeCol} / 1000), "%Y-%M") as group_key,
+    COUNT(${sortCol}) as group_value
+  FROM
+  (
+    SELECT ${sortCol} ${sortCol !== timeCol ? `, ${timeCol}` : ''}
+    FROM ${table}
+    ORDER BY ${sortCol}
```

Comment:
Not sure if it's intentional but this will give us the $LIMIT number of the oldest records, and since we are not pruning at the moment, that could give us the same results every time.

Reply:
Generally the sort column will be the row's id, so it shouldn't be tied to the time stamp / age of the record. I just noticed two spots where I don't use the row id, and I should. I'll fix this accordingly. Thanks for the call out.